File: src/rs/texture/SpriteTextureLoader.ts
import { Result, err, ok } from "../../util/Result";
import { DecodeError, decodeFailedError, notFoundError } from "../errors/DecodeError";
import { ByteBuffer } from "../io/ByteBuffer";
import { BytesProvider, EnumeratingBytesProvider } from "../io/BytesProvider";
import { IndexedSprite } from "../sprite/IndexedSprite";
import { SpriteLoader } from "../sprite/SpriteLoader";
import { brightenRgb } from "../util/ColorUtil";
import { TextureLoader } from "./TextureLoader";
import { TextureMaterial } from "./TextureMaterial";

export class SpriteTextureLoader implements TextureLoader {
    static readonly ANIM_DIRECTION_UV = [
        [0.0, 0.0],
        [0.0, -1.0],
        [-1.0, 0.0],
        [0.0, 1.0],
        [1.0, 0.0],
    ];

    idIndexMap: Map<number, number>;
    private readonly errors: Map<number, DecodeError>;

    static create(
        textureDefinitionSource: EnumeratingBytesProvider | undefined,
        spriteSource: BytesProvider,
    ): SpriteTextureLoader {
        const definitions = new Map<number, TextureDefinition>();
        const errors = new Map<number, DecodeError>();

        if (!textureDefinitionSource) {
            return new SpriteTextureLoader(spriteSource, [], definitions, errors);
        }
        const textureIds = Array.from(textureDefinitionSource.getIds());
        for (let i = 0; i < textureIds.length; i++) {
            const textureId = textureIds[i];
            const bytes = textureDefinitionSource.getBytes(textureId);
            if (!bytes) {
                continue;
            }
            const buffer = new ByteBuffer(bytes);
            const definitionResult = TextureDefinition.tryDecode(textureId, buffer);
            if (definitionResult.ok) {
                const definition = definitionResult.value;
                definitions.set(textureId, definition);
            } else {
                errors.set(textureId, definitionResult.error);
            }
        }

        return new SpriteTextureLoader(spriteSource, textureIds, definitions, errors);
    }

    constructor(
        readonly spriteSource: BytesProvider,
        readonly textureIds: number[],
        readonly definitions: Map<number, TextureDefinition>,
        errors?: Map<number, DecodeError>,
    ) {
        this.errors = errors ?? new Map();
        this.idIndexMap = new Map();
        for (let i = 0; i < textureIds.length; i++) {
            this.idIndexMap.set(textureIds[i], i);
        }
    }

    getTextureIds(): number[] {
        return this.textureIds;
    }

    getTextureIndex(id: number): number {
        return this.idIndexMap.get(id) ?? -1;
    }

    isSd(id: number): boolean {
        return true;
    }

    isSmall(id: number): boolean {
        const sprite = this.tryLoadTextureSprite(id);
        return sprite?.subWidth === 64;
    }

    getAverageHsl(id: number): number {
        return this.definitions.get(id)?.averageHsl ?? 0;
    }

    getAnimationUv(id: number): [number, number] {
        const def = this.definitions.get(id);
        if (!def) {
            return [0, 0];
        }

        const direction = def.animationDirection;
        const speed = def.animationSpeed;

        const uv = SpriteTextureLoader.ANIM_DIRECTION_UV[direction];

        return [uv[0] * speed, uv[1] * speed];
    }

    isTransparent(id: number): boolean {
        const def = this.definitions.get(id);
        if (!def) {
            return false;
        }
        return !def.opaque;
    }

    getMaterial(id: number): TextureMaterial {
        const def = this.definitions.get(id);
        if (!def) {
            return {
                animU: 0,
                animV: 0,
                alphaCutOff: 0.1,
            };
        }

        const direction = def.animationDirection;
        const speed = def.animationSpeed;

        const uv = SpriteTextureLoader.ANIM_DIRECTION_UV[direction];

        const animU = uv[0] * speed;
        const animV = uv[1] * speed;

        let alphaCutOff = 0.5;
        if (animU !== 0 || animV !== 0) {
            alphaCutOff = 0.1;
        }

        return {
            animU,
            animV,
            alphaCutOff,
        };
    }

    tryGetMaterial(id: number): TextureMaterial | undefined {
        const def = this.definitions.get(id);
        if (!def) {
            return undefined;
        }
        return this.getMaterial(id);
    }

    tryLoadMaterial(id: number): Result<TextureMaterial, DecodeError> {
        const def = this.definitions.get(id);
        if (!def) {
            return err(notFoundError("SpriteTextureMaterial", id));
        }
        return ok(this.getMaterial(id));
    }

    tryLoadTextureSprite(id: number): IndexedSprite | undefined {
        const result = this.tryLoadTextureSpriteResult(id);
        return result.ok ? result.value : undefined;
    }

    private tryLoadTextureSpriteResult(id: number): Result<IndexedSprite, DecodeError> {
        const cachedError = this.errors.get(id);
        if (cachedError) {
            return err(cachedError);
        }

        const def = this.definitions.get(id);
        if (!def) {
            const e = notFoundError("SpriteTextureDefinition", id);
            this.errors.set(id, e);
            return err(e);
        }

        for (let i = 0; i < def.spriteIds.length; i++) {
            const spriteId = def.spriteIds[i];
            const sprite = SpriteLoader.loadIntoIndexedSpriteFromSource(
                this.spriteSource,
                spriteId,
            );
            if (!sprite) {
                const e = decodeFailedError({
                    typeName: "SpriteTextureSprite",
                    id,
                    message: `SpriteTextureLoader: missing sprite for texture id=${id} spriteId=${spriteId}`,
                });
                this.errors.set(id, e);
                return err(e);
            }
            sprite.normalize();
            return ok(sprite);
        }

        const e = decodeFailedError({
            typeName: "SpriteTextureSprite",
            id,
            message: `SpriteTextureLoader: no sprites referenced by texture id=${id}`,
        });
        this.errors.set(id, e);
        return err(e);
    }

    private tryLoadPixelsInternal(
        id: number,
        size: number,
        flipH: boolean,
        brightness: number,
    ): Result<Int32Array, DecodeError> {
        const cachedError = this.errors.get(id);
        if (cachedError) {
            return err(cachedError);
        }

        const def = this.definitions.get(id);
        if (!def) {
            const e = notFoundError("SpriteTexturePixels", id);
            this.errors.set(id, e);
            return err(e);
        }

        const pixelCount = size * size;
        const pixels = new Int32Array(pixelCount);

        for (let i = 0; i < def.spriteIds.length; i++) {
            const spriteId = def.spriteIds[i];
            const sprite = SpriteLoader.loadIntoIndexedSpriteFromSource(
                this.spriteSource,
                spriteId,
            );
            if (!sprite) {
                const e = decodeFailedError({
                    typeName: "SpriteTexturePixels",
                    id,
                    message: `SpriteTextureLoader: missing sprite for texture id=${id} spriteId=${spriteId}`,
                });
                this.errors.set(id, e);
                return err(e);
            }
            sprite.normalize();

            const palettePixels = sprite.pixels;
            const transform = def.transforms[i];

            let index = 0;
            if (i > 0 && def.spriteTypes) {
                index = def.spriteTypes[i - 1];
            }

            if (index === 0) {
                const sourcePalette = sprite.palette;
                const paletteArgb = new Int32Array(sourcePalette.length);

                // not used by any texture but who knows
                const hasTransform = (transform & -0x1000000) === 0x3000000;
                const r_b = transform & 0xff00ff;
                const green = (transform >> 8) & 0xff;

                for (let pi = 0; pi < sourcePalette.length; pi++) {
                    let rgb = sourcePalette[pi];

                    if (hasTransform) {
                        const rg = rgb >> 8;
                        const gb = rgb & 0xffff;
                        if (rg === gb) {
                            const blue = rgb & 0xff;
                            const rbScaled = r_b * blue;
                            const gScaled = green * blue;
                            rgb = ((rbScaled >> 8) & 0xff00ff) | (gScaled & 0xff00);
                        }
                    }

                    let alpha = 0xff;
                    if (rgb === 0) {
                        alpha = 0;
                    }
                    paletteArgb[pi] = (alpha << 24) | brightenRgb(rgb, brightness);
                }

                if (size === sprite.subWidth) {
                    for (let pixelIndex = 0; pixelIndex < pixelCount; pixelIndex++) {
                        const paletteIndex = palettePixels[pixelIndex];
                        pixels[pixelIndex] = paletteArgb[paletteIndex];
                    }
                } else if (sprite.subWidth === 64 && size === 128) {
                    let pixelIndex = 0;

                    for (let x = 0; x < size; x++) {
                        for (let y = 0; y < size; y++) {
                            const paletteIndex = palettePixels[((x >> 1) << 6) + (y >> 1)];
                            pixels[pixelIndex++] = paletteArgb[paletteIndex];
                        }
                    }
                } else {
                    if (sprite.subWidth !== 128 || size !== 64) {
                        const e = decodeFailedError({
                            typeName: "SpriteTexturePixels",
                            id,
                            message: `SpriteTextureLoader: unsupported sprite size for texture id=${id} sprite.subWidth=${sprite.subWidth} size=${size}`,
                        });
                        this.errors.set(id, e);
                        return err(e);
                    }

                    let pixelIndex = 0;

                    for (let x = 0; x < size; x++) {
                        for (let y = 0; y < size; y++) {
                            const paletteIndex = palettePixels[(y << 1) + ((x << 1) << 7)];
                            pixels[pixelIndex++] = paletteArgb[paletteIndex];
                        }
                    }
                }
            }
        }

        return ok(pixels);
    }

    tryGetPixelsRgb(
        id: number,
        size: number,
        flipH: boolean,
        brightness: number,
    ): Int32Array | undefined {
        const result = this.tryLoadPixelsRgb(id, size, flipH, brightness);
        return result.ok ? result.value : undefined;
    }

    tryGetPixelsArgb(
        id: number,
        size: number,
        flipH: boolean,
        brightness: number,
    ): Int32Array | undefined {
        const result = this.tryLoadPixelsArgb(id, size, flipH, brightness);
        return result.ok ? result.value : undefined;
    }

    tryLoadPixelsRgb(
        id: number,
        size: number,
        flipH: boolean,
        brightness: number,
    ): Result<Int32Array, DecodeError> {
        return this.tryLoadPixelsInternal(id, size, flipH, brightness);
    }

    tryLoadPixelsArgb(
        id: number,
        size: number,
        flipH: boolean,
        brightness: number,
    ): Result<Int32Array, DecodeError> {
        return this.tryLoadPixelsInternal(id, size, flipH, brightness);
    }

    clearCache(): void {
        this.errors.clear();
    }
}

class TextureDefinition {
    static tryDecode(id: number, buffer: ByteBuffer): Result<TextureDefinition, DecodeError> {
        const averageHsl = buffer.readUnsignedShort();
        const opaque = buffer.readUnsignedByte() === 1;
        const spriteCount = buffer.readUnsignedByte();
        if (spriteCount < 1 || spriteCount > 4) {
            return err(
                decodeFailedError({
                    typeName: "SpriteTextureDefinition",
                    id,
                    message: `SpriteTextureLoader: invalid spriteCount=${spriteCount} for texture id=${id}`,
                }),
            );
        }

        const spriteIds = Array.from({ length: spriteCount }, () => buffer.readUnsignedShort());

        let spriteTypes: number[] | undefined;
        if (spriteCount > 1) {
            spriteTypes = Array.from({ length: spriteCount - 1 }, () => buffer.readUnsignedByte());
        }
        let unused: number[] | undefined;
        if (spriteCount > 1) {
            unused = Array.from({ length: spriteCount - 1 }, () => buffer.readUnsignedByte());
        }

        const transforms = Array.from({ length: spriteCount }, () => buffer.readInt());

        const animationDirection = buffer.readUnsignedByte();
        const animationSpeed = buffer.readUnsignedByte();

        return ok(
            new TextureDefinition(
                id,
                averageHsl,
                opaque,
                spriteCount,
                spriteIds,
                transforms,
                animationDirection,
                animationSpeed,
                spriteTypes,
                unused,
            ),
        );
    }

    constructor(
        readonly id: number,
        readonly averageHsl: number,
        readonly opaque: boolean,
        readonly spriteCount: number,
        readonly spriteIds: number[],
        readonly transforms: number[],
        readonly animationDirection: number,
        readonly animationSpeed: number,
        readonly spriteTypes?: number[],
        readonly unused?: number[],
    ) {}
}

File: cpp/rs/texture/SpriteTextureLoader.cpp
#include "SpriteTextureLoader.hpp"

#include <cstddef>

#include "../cache/format/Archive.hpp"
#include "../cache/format/ArchiveFile.hpp"
#include "../core/Allocator.hpp"
#include "../core/Move.hpp"
#include "../core/Result.hpp"
#include "../core/Span.hpp"
#include "../core/Status.hpp"
#include "../core/Vec.hpp"
#include "../io/Uint8ArrayReader.hpp"
#include "../sprite/SpriteLoader.hpp"
#include "../util/ColorUtil.hpp"
#include "../types.hpp"

namespace rs {

Status SpriteTextureDefinition::decode(i32 id, Span<const u8> bytes, SpriteTextureDefinition* out) noexcept {
    if (!out) {
        return Status::InvalidArgument;
    }
    Uint8ArrayReader reader(bytes, 0);

    u16 avg = 0;
    Status s = reader.readUnsignedShort(&avg);
    if (!ok(s)) {
        return s;
    }
    u8 opaqueU8 = 0;
    s = reader.readUnsignedByte(&opaqueU8);
    if (!ok(s)) {
        return s;
    }

    u8 spriteCount = 0;
    s = reader.readUnsignedByte(&spriteCount);
    if (!ok(s)) {
        return s;
    }
    if (spriteCount < 1 || spriteCount > 4) {
        return Status::BadFormat;
    }

    SpriteTextureDefinition def{};
    def.id = id;
    def.averageHsl = static_cast<i32>(avg);
    def.opaque = (opaqueU8 == 1);
    def.spriteCount = spriteCount;

    for (u8 i = 0; i < spriteCount; i++) {
        u16 spriteId = 0;
        s = reader.readUnsignedShort(&spriteId);
        if (!ok(s)) {
            return s;
        }
        def.spriteIds[i] = static_cast<i32>(spriteId);
    }

    if (spriteCount > 1) {
        def.hasSpriteTypes = true;
        const u8 typeCount = static_cast<u8>(spriteCount - 1);
        if (typeCount > 3) {
            return Status::BadFormat;
        }
        for (u8 i = 0; i < typeCount; i++) {
            u8 v = 0;
            s = reader.readUnsignedByte(&v);
            if (!ok(s)) {
                return s;
            }
            def.spriteTypes[static_cast<std::size_t>(i)] = v;
        }
        // Skip "unused" (spriteCount-1 bytes).
        for (u8 i = 0; i < typeCount; i++) {
            u8 v = 0;
            s = reader.readUnsignedByte(&v);
            if (!ok(s)) {
                return s;
            }
        }
    }

    for (u8 i = 0; i < spriteCount; i++) {
        i32 t = 0;
        s = reader.readInt(&t);
        if (!ok(s)) {
            return s;
        }
        def.transforms[i] = t;
    }

    u8 animDir = 0;
    u8 animSpeed = 0;
    s = reader.readUnsignedByte(&animDir);
    if (!ok(s)) {
        return s;
    }
    s = reader.readUnsignedByte(&animSpeed);
    if (!ok(s)) {
        return s;
    }
    def.animationDirection = animDir;
    def.animationSpeed = animSpeed;

    *out = def;
    return Status::Ok;
}

Result<SpriteTextureLoader> SpriteTextureLoader::create(const Archive& definitionArchive, const CacheIndex& spriteIndex, Allocator& alloc) noexcept {
    const Span<const ArchiveFile> files = definitionArchive.files();
    i32 maxId = -1;
    for (std::size_t i = 0; i < files.size(); i++) {
        if (files[i].id > maxId) {
            maxId = files[i].id;
        }
    }
    const i32 count = (maxId < 0) ? 0 : (maxId + 1);

    Vec<SpriteTextureDefinition> defs(alloc);
    auto rr = defs.resize(static_cast<std::size_t>(count));
    if (!rr.isOk()) {
        return Result<SpriteTextureLoader>::err(rr.status());
    }

    Vec<Status> statusById(alloc);
    rr = statusById.resize(static_cast<std::size_t>(count));
    if (!rr.isOk()) {
        return Result<SpriteTextureLoader>::err(rr.status());
    }
    for (std::size_t i = 0; i < statusById.size(); i++) {
        statusById[i] = Status::NotFound;
    }

    for (std::size_t i = 0; i < files.size(); i++) {
        const ArchiveFile& f = files[i];
        if (f.id < 0 || f.id >= count) {
            continue;
        }
        const i32 id = f.id;
        SpriteTextureDefinition def{};
        const Status s = SpriteTextureDefinition::decode(id, f.data.span(), &def);
        defs[static_cast<std::size_t>(id)] = def;
        statusById[static_cast<std::size_t>(id)] = s;
    }

    return Result<SpriteTextureLoader>::ok(SpriteTextureLoader(&spriteIndex, rs::move(defs), rs::move(statusById), count));
}

Result<SpriteTextureLoader> SpriteTextureLoader::createEmpty(const CacheIndex& spriteIndex, Allocator& alloc) noexcept {
    Vec<SpriteTextureDefinition> defs(alloc);
    Vec<Status> statusById(alloc);
    return Result<SpriteTextureLoader>::ok(SpriteTextureLoader(&spriteIndex, rs::move(defs), rs::move(statusById), 0));
}

Status SpriteTextureLoader::getDefinition(i32 id, const SpriteTextureDefinition** out) const noexcept {
    if (!out) {
        return Status::InvalidArgument;
    }
    if (id < 0) {
        *out = nullptr;
        return Status::OutOfRange;
    }
    const std::size_t idx = static_cast<std::size_t>(id);
    if (idx >= statusById_.size()) {
        *out = nullptr;
        return Status::OutOfRange;
    }
    const Status s = statusById_[idx];
    if (!ok(s)) {
        *out = nullptr;
        return s;
    }
    *out = &defs_[idx];
    return Status::Ok;
}

Result<Vec<i32>> SpriteTextureLoader::tryGetPixelsArgb(i32 id, i32 size, bool flipH, float brightness, Allocator& alloc) const noexcept {
    (void)flipH;

    if (!spriteIndex_) {
        return Result<Vec<i32>>::err(Status::InvalidArgument);
    }

    const SpriteTextureDefinition* def = nullptr;
    Status s = getDefinition(id, &def);
    if (!ok(s) || !def) {
        return Result<Vec<i32>>::err(s);
    }

    if (size <= 0) {
        return Result<Vec<i32>>::err(Status::InvalidArgument);
    }
    const std::size_t pixelCount = static_cast<std::size_t>(size) * static_cast<std::size_t>(size);

    Vec<i32> pixels(alloc);
    auto rr = pixels.resize(pixelCount);
    if (!rr.isOk()) {
        return Result<Vec<i32>>::err(rr.status());
    }
    for (std::size_t i = 0; i < pixels.size(); i++) {
        pixels[i] = 0;
    }

    for (u8 spriteIdx = 0; spriteIdx < def->spriteCount; spriteIdx++) {
        const i32 spriteId = def->spriteIds[spriteIdx];

        // Fetch sprite bytes from sprites index: archiveId=spriteId, fileId=0.
        auto archRes = spriteIndex_->getArchive(spriteId, alloc);
        if (!archRes.isOk()) {
            return Result<Vec<i32>>::err(archRes.status());
        }
        const Archive spriteArchive = rs::move(archRes.value());
        const ArchiveFile* file = spriteArchive.getFile(0);
        if (!file) {
            return Result<Vec<i32>>::err(Status::NotFound);
        }

        auto spriteRes = SpriteLoader::decodeSpriteArchive(file->data.span(), alloc);
        if (!spriteRes.isOk()) {
            return Result<Vec<i32>>::err(spriteRes.status());
        }
        SpriteArchive decoded = rs::move(spriteRes.value());
        if (decoded.sprites.size() == 0) {
            return Result<Vec<i32>>::err(Status::NotFound);
        }

        IndexedSprite& sprite = decoded.sprites[0];
        s = sprite.normalize(alloc);
        if (!ok(s)) {
            return Result<Vec<i32>>::err(s);
        }

        const Span<const i32> sourcePalette = sprite.palette;
        Vec<i32> paletteArgb(alloc);
        rr = paletteArgb.resize(sourcePalette.size());
        if (!rr.isOk()) {
            return Result<Vec<i32>>::err(rr.status());
        }

        const i32 transform = def->transforms[spriteIdx];
        const bool hasTransform = (transform & -0x01000000) == 0x03000000;
        const i32 r_b = transform & 0x00FF00FF;
        const i32 green = (transform >> 8) & 0xFF;

        for (std::size_t pi = 0; pi < sourcePalette.size(); pi++) {
            i32 rgb = sourcePalette[pi];
            if (hasTransform) {
                const i32 rg = rgb >> 8;
                const i32 gb = rgb & 0xFFFF;
                if (rg == gb) {
                    const i32 blue = rgb & 0xFF;
                    rgb = (((r_b * blue) >> 8) & 0x00FF00FF) | ((green * blue) & 0x0000FF00);
                }
            }

            i32 alpha = 0xFF;
            if (rgb == 0) {
                alpha = 0;
            }
            const i32 bright = brightenRgb(rgb, brightness);
            paletteArgb[pi] = (alpha << 24) | bright;
        }

        const i32 sw = sprite.subWidth;
        const i32 sh = sprite.subHeight;
        const Span<const u8> palettePixels(sprite.pixels.data(), sprite.pixels.size());

        std::size_t outIndex = 0;
        if (size == sw) {
            for (std::size_t p = 0; p < pixelCount; p++) {
                const u8 idx = palettePixels[p];
                pixels[p] = paletteArgb[static_cast<std::size_t>(idx)];
            }
            continue;
        }

        if (sw == 64 && size == 128) {
            for (i32 x = 0; x < size; x++) {
                for (i32 y = 0; y < size; y++) {
                    const std::size_t src = (static_cast<std::size_t>(x >> 1) << 6) + static_cast<std::size_t>(y >> 1);
                    const u8 idx = palettePixels[src];
                    pixels[outIndex++] = paletteArgb[static_cast<std::size_t>(idx)];
                }
            }
            continue;
        }

        if (sw == 128 && size == 64) {
            for (i32 x = 0; x < size; x++) {
                for (i32 y = 0; y < size; y++) {
                    const std::size_t src =
                        static_cast<std::size_t>(y << 1) + ((static_cast<std::size_t>(x << 1)) << 7);
                    const u8 idx = palettePixels[src];
                    pixels[outIndex++] = paletteArgb[static_cast<std::size_t>(idx)];
                }
            }
            continue;
        }

        return Result<Vec<i32>>::err(Status::Unsupported);
    }

    return Result<Vec<i32>>::ok(rs::move(pixels));
}

} // namespace rs
